joblib/numpy_pickle.py

```diff
@@ -296,37 +573,48 @@ class NumpyUnpickler(Unpickler):
     dispatch = Unpickler.dispatch.copy()
 
     def __init__(self, filename, file_handle, mmap_mode=None):
+        # The 2 next lines are for backward compatibility
         self._filename = os.path.basename(filename)
         self._dirname = os.path.dirname(filename)
+
         self.mmap_mode = mmap_mode
-        self.file_handle = self._open_pickle(file_handle)
+        self.file_handle = file_handle
+        self.filename = filename
         Unpickler.__init__(self, self.file_handle)
         try:
             import numpy as np
         except ImportError:
             np = None
         self.np = np
-
-    def _open_pickle(self, file_handle):
-        return file_handle
+        self.current_offset = c_int64(-1)
 
     def load_build(self):
         """ This method is called to set the state of a newly created
             object.
-
             We capture it to replace our place-holder objects,
             NDArrayWrapper, by the array we are interested in. We
             replace them directly in the stack of pickler.
         """
         Unpickler.load_build(self)
+
+        # For back backward compatibility
         if isinstance(self.stack[-1], NDArrayWrapper):
             if self.np is None:
-                raise ImportError('Trying to unpickle an ndarray, '
-                        "but numpy didn't import correctly")
+                raise ImportError("Trying to unpickle an ndarray, "
+                                  "but numpy didn't import correctly")
             nd_array_wrapper = self.stack.pop()
             array = nd_array_wrapper.read(self)
             self.stack.append(array)
 
+        if isinstance(self.stack[-1], NPArrayWrapper):
+            if self.np is None:
+                raise ImportError("Trying to unpickle an ndarray, "
+                                  "but numpy didn't import correctly")
+            np_array_wrapper = self.stack.pop()
+            array = np_array_wrapper.read(self)
+            # push back the reconstructed on the unpickler stack
+            self.stack.append(array)
```

Comment:
This code is duplicated from above you could tackle both exception types using a tuple as the isinstance argument:

```python
if isinstance(self.stack[-1], (NDArrayWrapper, NPArrayWrapper)):
```

Not sure about that, but if we want to get rid of the deprecated pickle format one day, maybe we want to have a deprecation warning for the NDArrayWrapper case, asking the user nicely to regenerate his pickle. This could get very quickly annoying for old joblib caches though.